### PR TITLE
Add codebase doc (ECOOP23 artifact version)

### PR DESCRIPTION
**Pull request by andongfan**

### CODE REVIEW

File: doc/mls-codebase-doc.md [hunk 1]
```diff
@@ -0,0 +1,354 @@
+# Document to the Codebase of the MLscript Compiler
```
Comment (LPTK) [inline doc/mls-codebase-doc.md]: It's *documentation*, not *document*... Please also correct below.

File: doc/mls-codebase-doc.md [hunk 2]
```diff
@@ -0,0 +1,354 @@
...
+
+This is the document to the codebase of the MLscript Compiler for the reuse
+of our MLscript implementation.
```
Comment (LPTK) [inline doc/mls-codebase-doc.md]:
```suggestion
This is the documentation of the MLscript codebase.
```

**Review by LPTK — COMMENTED** [review]
Thanks!

But after I started reading it appeared to me you haven't checked the grammar. Please make a thorough Grammarly pass and make sure everything is clearly written.

**Comment by LPTK** [comment]
@andongfan the PR was closed inadvertently (and silently! – GitHub didn't notify me) because I deleted the branch on my fork... Please when you have a moment get back to this so people can benefit from it.

**Comment by andongfan** [comment]
Sorry for my delay. Synced with the documentation in the published artifact.

**Comment by LPTK** [comment]
Please read https://github.com/hkust-taco/mlscript/blob/mlscript/CONTRIBUTING.md first.

It seems you didn't even read my comments above, nor did you take an overall look at your diff.

Please request my review when you have done so.

**Comment by andongfan** [comment]
I made a thorough pass on the whole documentation (except for the codegen section) to improve language and tried to fix obvious inconsistencies with the latest codebase as much as I could for now. I found the code organization of inheritance has changed a lot since the artifact, but I believe the idea is the same.

I'm asking Cunyuan to take a look at the codegen part since he composed that part.

@LPTK Your prior suggestions have been applied. Sorry for overlooking them.

**Review by LPTK — APPROVED** [review]
Ok it's looking pretty good now. Thanks guys!

**Pull request merged**